txnsync/encodedgroupsunmarshalers.go

```diff
@@ -0,0 +1,795 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package txnsync
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/transactions"
+)
+
+var errDataMissing = errors.New("failed to decode: data missing")
+
+func getSlice(b []byte, index int, size int) ([]byte, error) {
+	if index*size+size > len(b) {
+		return nil, errDataMissing
+	}
+	return b[index*size : index*size+size], nil
+}
+
+func getNibble(b []byte, index int) (byte, error) {
+	if index > len(b)*2 {
+		return 0, errDataMissing
+	}
+	if index%2 == 0 {
+		return b[index/2] / 16, nil
+	}
+	return b[index/2] % 16, nil
+}
+
+func (stub *txGroupsEncodingStub) reconstructSignedTransactions(signedTxns []transactions.SignedTxn, genesisID string, genesisHash crypto.Digest) error {
+	var index int
+	index = 0
+	for i := range signedTxns {
+		if exists := stub.BitmaskSig.EntryExists(i, int(stub.TotalTransactionsCount)); exists {
```

Comment:
looks as if you could have reduced this to 
```golang
if stub.BitmaskSig.EntryExists(i, int(stub.TotalTransactionsCount)) {
```
and make it (slightly) simpler.

Reply:
Another thing - I see that in all the calls to `EntryExists` you're casting the second parameter to int. It might be a good indicator that either this parameter should have been an int, or that the method parameter should have been different.

Reply:
The main problem is that I'd like `stub.TotalTransactionsCount` to be an int and msgp doesn't let me

Reply:
The method parameter doesn't make much sense to be a uint64